db/snapshot/backup_task.cc

```diff
@@ -103,52 +105,89 @@ future<> backup_task_impl::upload_component(shared_ptr<s3::client>& client, sstr
-        throw std::invalid_argument(fmt::format("snapshot does not exist at {}", _snapshot_dir.native()));
-    }
+namespace {
 
+future<std::vector<std::tuple<size_t, sstring>>> get_backup_files(const std::filesystem::path& snapshot_dir) {
     std::exception_ptr ex;
-    gate uploads;
-    auto snapshot_dir_lister = directory_lister(_snapshot_dir, lister::dir_entry_types::of<directory_entry_type::regular>());
-    auto cln = _snap_ctl.storage_manager().container().local().get_endpoint_client(_endpoint);
+    auto snapshot_dir_lister = directory_lister(snapshot_dir, lister::dir_entry_types::of<directory_entry_type::regular>());
+    std::vector<std::tuple<size_t, sstring>> backup_files;
+    std::optional<directory_entry> component_ent;
 
-    for (;;) {
-        std::optional<directory_entry> component_ent;
+    do {
         try {
             component_ent = co_await snapshot_dir_lister.get();
-        } catch (...) {
-            if (!ex) {
-                ex = std::current_exception();
-                break;
+            if (component_ent) {
+                auto component_name = snapshot_dir / component_ent->name;
+                auto size = co_await file_size(component_name.native());
+                backup_files.emplace_back(size, std::move(component_ent->name));
             }
-        }
-        if (!component_ent.has_value()) {
+        } catch (...) {
+            ex = std::current_exception();
             break;
         }
-        auto gh = uploads.hold();
+    } while (component_ent);
+    co_await snapshot_dir_lister.close();
+    if (ex)
+        co_await coroutine::return_exception_ptr(std::move(ex));
 
-        // Pre-upload break point. For testing abort in actual s3 client usage.
-        co_await utils::get_local_injector().inject("backup_task_pre_upload", utils::wait_for_message(std::chrono::minutes(2)));
+    std::ranges::sort(backup_files, std::greater());
+    co_return backup_files;
+}
+} // namespace
 
-        std::ignore = upload_component(cln, component_ent->name).handle_exception([&ex] (std::exception_ptr e) {
-            // keep the first exception
-            if (!ex) {
-                ex = std::move(e);
-            }
-        }).finally([gh = std::move(gh)] {});
-        co_await coroutine::maybe_yield();
-        co_await utils::get_local_injector().inject("backup_task_pause", utils::wait_for_message(std::chrono::minutes(2)));
-        if (impl::_as.abort_requested()) {
-            ex = impl::_as.abort_requested_exception_ptr();
-            break;
-        }
+future<> backup_task_impl::do_backup() {
+    if (!co_await file_exists(_snapshot_dir.native())) {
+        throw std::invalid_argument(fmt::format("snapshot does not exist at {}", _snapshot_dir.native()));
     }
 
-    co_await snapshot_dir_lister.close();
-    co_await uploads.close();
-    if (ex) {
-        co_await coroutine::return_exception_ptr(std::move(ex));
+    auto backup_files = co_await get_backup_files(_snapshot_dir);
```

Comment:
It's still "static" partitioning of files between shards, so the main concern is still not addressed -- in case one shard is slow (e.g. compaction kicks in and steals CPU and disk IO) and other shards have more resources, they will finish upload and will wait for this slow one

Reply:
Not sure I'm following. It is just a list of backup fileshow it supposed to address anything? each shard just pickups a file from it whenever it is convinient for the shard

Reply:
[lines 165-171](https://github.com/scylladb/scylladb/pull/22634/files#diff-483d9a13ecdedf71bd4c561af0b32e33386ffbfc4273dc8f482ebb0e78602539R165-R171) indicates that indeed this list is being "eaten" by whoever is free to collect, no waits on slowed shards.